lib/decompress/zstd_decompress_block.c

```diff
@@ -1173,9 +1178,10 @@ ZSTD_decompressSequences_body( ZSTD_DCtx* dctx,
         __asm__(".p2align 4");
 #  endif
 #endif
+        /* decompress without overrunning litPtr begins */
         for ( ; ; ) {
             seq_t const sequence = ZSTD_decodeSequence(&seqState, isLongOffset);
-            size_t const oneSeqSize = ZSTD_execSequence(op, oend, sequence, &litPtr, litEnd, prefixStart, vBase, dictEnd);
+            size_t const oneSeqSize = ZSTD_execSequence(op, oend, litPtr + sequence.litLength - WILDCOPY_OVERLENGTH, sequence, &litPtr, litEnd, prefixStart, vBase, dictEnd);
```

Comment:
I think that `oend_w = litPtr + sequence.litLength - WILDCOPY_OVERLENGTH` isn't quite right. It works when the `litLength` is "valid". But, at this point we haven't yet verified that `litLength` is valid. If it is corrupt, or the output buffer is too small, we could end up with `oend_w > oend`, which might allow for out of bounds writes.

This command will help you fuzz test on invalid inputs:

```
cd tests/fuzz
# Download the fuzz corpora
make -j corpora
# Build the simple_decompress fuzzer
CC=clang CXX=clang++ ./fuzz.py build all --enable-asan --enable-ubsan --enable-fuzzer
# List the fuzzers
./fuzz.py list
# Run the simple_decompress fuzzer
./fuzz.py libfuzzer simple_decompress
# Run the stream_decompress fuzzer
./fuzz.py libfuzzer stream_decompress

```

Reply:
Those commands are testing decompression of invalid inputs. You can also test round trips with the `simple_round_trip`, and `stream_round_trip` fuzzers.

When making a complex change like this, I recommend running the relevant fuzzers before landing, to catch any low hanging fruit. I'll generally run them for ~1 hour. But, these fuzzers are also run continuously by oss-fuzz after landing, and that will catch rarer bugs that might not be caught within 1 hour. And we're okay with the occasional hard to trigger bug reaching the `dev` branch, as long as our fuzzers catch it before release.